### ISSUE TITLE
BUDGET: Create class method for determining money pledged in a given week

### ISSUE THREAD

**Issue by colinxfleming** [colinxfleming/dcaf_case_management#709]
Our goal here: Take all patients whose appointment date is in a given week and calculate the DCAF pledge amount. Here's the rub: I don't know what the best way to do it is. 

Here are the requirements: We need to be able to filter to pledges in a given calendar week, and distinguish those between pledges that have been sent and those that haven't.

Here's a suggested path to this, but I'm amenable to another option:
* Create a class method on the `Patient` object that filters down to anyone with an appointment date in a given week and maps that collection of objects down to a hash like `{ pledged: 1000, sent: 500 }`.

We'll in turn use this to build a helper method that takes a hash as an input and spits out a nice looking budget bar.

**Comment by lomky** [comment]
I'll take a stab at this :grinning: 

My only question is if it would be one function or two:

Are there other use cases that would want all the patients with appointments in the next week?
 - if so, then we could teach the `Patient` object a more generic function like `patients_with_upcoming_appointment` taking a duration parameter, and then maybe move the data extraction we're looking for into a helper.

Alternately if that is over-engineering it, we could always make that split later if we ran into another use case later.

**Comment by colinxfleming** [comment]
Love the idea of abstracting it like that -- if this is something that
lends itself well to that, I'd say go for it!

On Thu, Nov 3, 2016 at 7:38 PM Kathryn Tipton <notifications@github.com>
wrote:

I'll take a stab at this 😀

My only question is if it would be one function or two:

Are there other use cases that would want all the patients with
appointments in the next week?

   - if so, then we could teach the Patient object a more generic function
   like patients_with_upcoming_appointment taking a duration parameter, and
   then maybe move the data extraction we're looking for into a helper.

Alternately if that is over-engineering it, we could always make that split
later if we ran into another use case later.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/colinxfleming/dcaf_case_management/issues/709#issuecomment-258306246>,
or mute the thread

**Comment by lomky** [comment]
I'm not going to get this to a PR state tonight. I'll try to get it there this weekend!

**Comment by lomky** [comment]
@colinxfleming I think I have found where this is sticking in my head. I'm used to a slightly different MVC pattern.

I work mostly in Mojolicious, which has distinct model controllers for dealing with single objects vs collections of objects. E.g. CRUD code dealing with a single Patient would go in `./Patient.pm` and CRUD code dealing with multiple Patients would go in `./Patient/Manager.pm`.

Where does ruby handle manipulating multiple objects of a type? Does this go in the model alongside single-object manipulation? Into the controller? Or since this particular function is enabling a shared functionality, does it go into a helper?

At this point I actually have the logic psuedo-coded, I just don't know where to put it. When I tried to slot it into the model, I realized I couldn't figure out how to write a test for it there.

Psuedocode:

```
outstanding_pledges = 0
sent_total = 0
Patient.where(appointment_date.lte => Date.today + 7).each do |patient|
  if patient.pregnancy.pledge_sent 
    sent_total += patient.pregnancy.dcaf_soft_pledge 
  else
    outstanding_pledges =  patient.pregnancy.dcaf_soft_pledge
  end
end
return { pledged: outstanding_pledges, sent: sent_total
```

**Comment by colinxfleming** [comment]
@lomky excellent questions, really appreciate that you're engaging with the rails 'style' here!

the short version is that code that deals with that object or objects of that type goes in the same class definition. You can check out some similar code under `Patient#urgent_patients` -- it's a similar hit of everyone in the db that just filters down to people who have the urgent flag set to true. If this were more of an API-oriented app instead we'd probably do this on the controller level somehow. Helpers are mostly for view-related code.

Here I think is a helpful-ish blog post that's an overview of similar patterns: http://culttt.com/2015/06/10/understanding-class-methods-verses-instance-methods-in-ruby/
. The short version is that methods defined with `def self.method` are available when you do `Patient.method`, but methods defined with a regular old `def method` are available only to individual objects (e.g. you have to retrieve an individual patient from the database to do that, such as in `Patient.first.method`).

re: your pseudocode, that pretty much looks right to me! Pretty much all you'll need to do is chuck that in `def self.whatever` -- the `Patient` on line three will also be inferred, so you should just be able to start it `where`. 

Let me know if that unsticks you -- this looks solid!

**Comment by lomky** [comment]
@colinxfleming Brilliant! Instance vs Class completely unstuck my head. I have been too long away from a proper OO language 🐫

**Comment by colinxfleming** [comment]
💯 💯 💯 💯 💯

**Comment by lomky** [comment]
I cannot seem to get the tests to pass. :( submitting a PR so others can review the code and hopefully help spot my bug!

**Comment by colinxfleming** [comment]
Deeply appreciate that you took the time to go the extra mile and write
tests! I'll have some feedback for you by the end of the day if not sooner.

**Comment by colinxfleming** [comment]
Fixed by @lomky in #717 !
